_posts/2018-12-06-pmreorder-basics.md

```diff
@@ -0,0 +1,318 @@
+	struct list_node *node2 = NODE_PTR(root, node1->next);
+	struct list_node *node3 = NODE_PTR(root, node2->next);
+
+	if ((root->head == 6 && node1->value != 66) || \
+		(node1->next == 3 && node2->value != 33) || \
+		(node2->next == 5 && node3->value != 55))
+		return 1;
+
+	return 0;
+}
+{% endhighlight %}
+
+That's it. List and consistency checker are ready.
+The next step is to generate a log of memory operations
+made by the application. Such a log we will call store log.
+
+
+### Store log
+
+To get the list of all memory operations we are using pmemcheck logging
+functionality. To turn on logging use `log-stores` parameter:
+
+{% highlight sh %}
+$ valgrind --tool=pmemcheck --log-stores=yes --log-file=store_log.log <your_app> [opt]
+{% endhighlight %}
+
+An output of the above command is a text file with the list of stores, flushes, fences,
+user-defined markers and possibly some other data that can be consumed by pmreorder tool.
+
+Let's look at part of such a store log, and read comments
+next to the lines of the log to distinguish stores:
+
+{% highlight sh %}
+STORE;0x5600060;0x0;0x8|	# set new->next for node_id 5
+FLUSH;0x5600040;0x40|		# persist next for node_id 5
+FENCE|
+STORE;0x5600000;0x5;0x8|	# set root->head for node_id 5
+FLUSH;0x5600000;0x40|		# persist head for node_id 5
+FENCE|
+STORE;0x5600058;0x37;0x4|	# set new->value for node_id 5
+FLUSH;0x5600040;0x40|		# persist value for node_id 5
+FENCE|
+STORE;0x5600040;0x5;0x8|	# set new->next for node_id 3
+FLUSH;0x5600040;0x40|		# persist  next for nod_id 3
+FENCE|
+STORE;0x5600000;0x3;0x8|	# set root->head for node_id 3
+FLUSH;0x5600000;0x40|		# persist head for node_id 3
+FENCE|
+STORE;0x5600038;0x21;0x4|	# set new->value for node_id 3
+FLUSH;0x5600000;0x40|		# persist value for node_id 3
+FENCE|
+STORE;0x5600070;0x3;0x8|	# set new->next for node_id 6
+FLUSH;0x5600040;0x40|		# persist next for node_id 6
+FENCE|
+STORE;0x5600000;0x6;0x8|	# set root->head for node_id 6
+FLUSH;0x5600000;0x40|		# persist head for node_id 6
+FENCE|
+STORE;0x5600068;0x42;0x4|	# set new->value for node_id 6
+FLUSH;0x5600040;0x40|		# persist value for node_id 6
+FENCE|
+STOP
+==32611== Number of stores not made persistent: 0
+==32611== ERROR SUMMARY: 0 errors
+{% endhighlight %}
+
+It often happends that store log contains multiple stores and operations
+made on the pool that are irrelevant for your tests.
+There is a mechanism called markers that allows ignoreing selected set of stores
+during consistency check. For more information about this one and others features
+look at [pmreorder man page][pmreorderManLink].
+
+As you probably guess, store log generation was the last step before
+pmreorder execution, so let's move on.
+
+
+### Pmreorder
+
+This probably will be the most important paragraph in this blog post,
+so let me start from a brief description of how pmreorder works.
+
+Pmreorder parses store log provided by the user and reproduce on
```

Comment:
and generates sequences of stores between barriers. Those sequences are written to a pool file. For each generated sequence, the tool runs the consistency function, and reports those scenario for which the function returned 1.